third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,512 @@
+/// where the `deref_mut` method simply returns a mutable reference to the wrapped value.
+/// Note that `Wrapper` should've implemented `Deref`, which is a supertrait of `DerefMut`.
+macro_rules! impl_deref_mut {
+    ($struct_name:ident) => {
+        impl std::ops::DerefMut for $struct_name {
+            fn deref_mut(&mut self) -> &mut Self::Target {
+                &mut self.0
+            }
+        }
+    };
+}
+
+impl_deref!(RemoveEmptyBlock, ControlFlowGraphCodeGenerator);
+
+impl_deref_mut!(RemoveEmptyBlock);
+
+impl RemoveEmptyBlock {
+    /// Wrapper
+    pub fn new(generator: ControlFlowGraphCodeGenerator) -> Self {
+        Self(generator)
+    }
+
+    /// - Eliminates branch/jump to jump
+    ///     L1: goto L2 (L1 != L2)
+    ///     =>
+    ///     (removed)
+    ///
+    ///     goto L1
+    ///     =>
+    ///     goto L2
+    ///
+    ///     if ... goto L1 else ... / if ... goto ... else L1
+    ///     =>
+    ///     if ... goto L2 else ... / if ... goto ... else L2
+    pub fn transform(&mut self) {
+        for block in self.cfg.blocks() {
+            if self.is_empty_block(block) {
+                let suc_block = self.get_the_non_trivial_successor(block);
+                if suc_block != block {
+                    self.remove_empty_block(block, suc_block);
+                }
+            }
+        }
+    }
+
+    /// Checks if the given block is empty; i.e., only a label and a jump
+    fn is_empty_block(&self, block_id: BlockId) -> bool {
+        let block_instrs = self.block_instrs(block_id);
+        block_instrs.len() == 2
+            && matches!(block_instrs[0], Bytecode::Label(..))
+            && matches!(
+                block_instrs.last().expect("instruction"),
+                Bytecode::Jump(..)
+            )
+    }
+
+    /// Removes the block from the control flow graph, and let any block jumpping to it
+    /// to jump to `redirect_to` directly
+    /// Requires: `block_to_remove` doesn't have itself as a successor. We don't remove `loop {}`.
+    fn remove_empty_block(&mut self, block_to_remove: BlockId, redirect_to: BlockId) {
+        debug_assert!(block_to_remove != redirect_to);
+        debug_assert!(!self
+            .cfg
+            .successors(block_to_remove)
+            .contains(&block_to_remove));
+        let maybe_preds = self.predecessors.remove(&block_to_remove);
+        if let Some(preds) = maybe_preds {
+            for pred in preds {
+                if pred != self.cfg.entry_block() {
+                    let from = self.get_block_label(block_to_remove).expect("label");
+                    let to = self.get_block_label(redirect_to).expect("label");
+                    let pred_codes = self.code_blocks.get_mut(&pred).expect("code block");
+                    Self::redirects_block(pred_codes, from, to);
+                }
+                // update successors of predecessors of `block_to_remove`
+                for suc_of_pred in self.cfg.successors_mut(pred) {
```

Comment:
This is weird. Why do we need to update the cfg? Why aren't all updates contained to the generator?

I see you added this `successors_mut` to the cfg, but it now opens the door to make the data structure inconsistent via outside mutation, and also makes it harder to change the internal representation. If you need to mutate do it on a new data structure? You already have `predecessors`. 

The other option is that all this becomes part of the graph module instead of being here, in which case it makes sense it can access internals.

Reply:
?

Reply:
done